enhancements/installer/azurestack.md

```diff
@@ -0,0 +1,213 @@
+			...
+		}.JSON()
+		if err != nil {
+			return errors.Wrap(err, "could not create cloud provider config endpoints")
+		}
+		cm.Data[cloudProviderEndpointsKey] = azureStackEndpoints
+
+```
+
+Operators can then [create clients](https://github.com/Azure/azure-sdk-for-go/blob/master/services/attestation/mgmt/2020-10-01/attestation/client.go#L45-L53) using the resource manager endpoint and subscription ID from the cloud provider
+configs.
+#### Installer Manifests & Cloud Provider Config
+ASH should be able to reuse Azure machine manifests with proper validation.
+
+The [Azure cloud-provider-config](https://github.com/openshift/installer/blob/master/pkg/asset/manifests/azure/cloudproviderconfig.go) will need to be slightly adapted. `LoadBalancerSku` for ASH will be `basic`. `UseInstanceMetadata` should probably be `false`.
+
+The cluster [DNS manifest contains the private DNS zone](https://github.com/openshift/installer/blob/master/pkg/asset/manifests/dns.go#L110-L112). At a minimum this should be removed, but there is an [open question](#open-question) about how this should be handled.
+
+### User Authentication
+
+Users of the installer will authenticate with service principals. The standard Azure workflow works here.
+
+### Lack of Private DNS Zones & Baremetal Networking
+ASH does not provide private DNS zones. The [on-prem, baremetal networking solution](../network/baremetal-networking.md) will be considered to provide private DNS. It remains to be seen whether this can be used with the Azure load balancers.
+
+### Install Config
+Users must provide a resource manager endpoint. A new field will need to be added to the install config for the URI.
+
+As discussed above, the endpoint follows a pattern: `https://management.{region}.{Azure Stack Hub domain}`. As the install config contains only the region, we could add a field for only the Azure Stack domain and construct the endpoint. On the other hand, I have seen an example where the endpoint did not follow this pattern (`adminmanagement` instead of
+`management`) but this was not a production environment so may not be an actual problem. For advanced use, this value could
+be edited in the manifests. If we believe having users supply only the domain is more ergonomic, then this does seem like
+a viable option that would capture most use cases.
+
+Another option to consider, would be adding the ability to [supply all endpoints, similar to AWS](https://github.com/openshift/installer/blob/master/pkg/types/aws/platform.go#L39).
+### Terraform
+Azure Stack has [a stand-alone Terraform provider](https://registry.terraform.io/providers/hashicorp/azurestack/latest/docs).
+
+This provider lacks the ability to create a service principal for the resource group and assign a contributor role, which is required by the Ingress controller.
+
+These actions are achievable through the CLI, if necessary these commands could be run as a Terraform post hook.
```

Comment:
Is this conflating the service principal used by the ingress operator with the user-assigned identity attached to the VMs, which is used by the kube-controller-manager? For public Azure, the terraform creates a user-assigned identity but does not create any service principals. Azure Stack does not support user-assigned identities, even through the CLI.

Reply:
My statement in the enhancement is confused. The issue is, as you say, the lack of user-assigned identities--not service principals, but it looks like the "main" identity is scoped to the resource group. Here is the code from [IPI](https://github.com/openshift/installer/blame/619d0b28f377ab0c397b1a8b169dac5429599a38/data/data/azure/main.tf#L96-L107). The  [UPI](https://github.com/openshift/installer/blob/master/docs/user/azure/install_upi.md#grant-access-to-the-identity) doc states:

> Grant the Contributor role to the Azure identity so that the Ingress Operator can create a public IP and its load balancer.

I'm not sure I have a firm grasp on the problem. From reading the docs it _seems_ we need to create this identity, to give operators access to create resources in the resource group. The machine API operator would probably need this as well, but perhaps it is not mentioned as these are UPI docs.

Really what I am missing is why/how isn't this handled by the CCO? If we're using passthrough I assume this isn't a problem at all because those creds clearly have perms to create resources in the resource group.

Let me know what you think and I'll update the doc accordingly.

Reply:
This is handled by the CCO. All of the in-cluster operators use credentials that are granted by the CCO (assuming mint mode). The UPI docs are a bit misleading when they say that the Contributor role is needed by the Ingress operator. The Ingress operator is creating loadbalancer Services, but it is the kube-controller-manager that is creating the cloud resources for those Services.

In public Azure, the kube-controller-manager uses the user-assigned identity assigned to the VM. Azure Stack does not support user-assigned identities. Presumably, we could allow Azure Stack to create system-assigned identities for each VM. The installer would then need to assign the Contributor role to each system-assigned identity **after** each VM was created. The kubelet also uses the managed identity assigned to the VM, so the managed identity is need for worker machines too. However, the installer does not create the worker VMs so cannot assign the Contributor role to the system-assigned identities for the worker VMs. As far as I am aware, the machine-api does not have a feature for assigning roles to the system-assigned identities of the VMs that the machine-api creates.

A possible solution is to add that feature to the machine-api. Another solution is to supply a Service Principal in the cloud config that the kubelet and kube-controller-manager would use when accessing the Azure API. The latter solution is what I think you are using currently in your UPI work. There may be problems with utilizing that solution long-term due to how the user would replace that Service Principal. I don't think that simply changing the Service Principal in the cloud-provider-config ConfigMap would cause the replacement Service Principal to roll out to all of the kubelets and kube-controller-managers.

Reply:
I get it now. The piece I was missing is that the main identity is assigned
to the VMs in TF and the machinesets. Then you explained the rest; thank
you for that!

If we want to manage the roles after creation, then VM extensions may
provide this:
https://registry.terraform.io/providers/hashicorp/azurestack/latest/docs/resources/virtual_machine_extension


But it seems the MAO may be more involved. Yes the UPI solution writes the
cloud provider config with the service principal ID & secret. So I guess
the question is: does updating the cloud provider config configmap in
cluster cause a new machine config to be written? Does a new machine config
cause a reboot, in which case the new creds would be picked up by the
kubelet.